src/app/utilities/moodle/helper.ts

```diff
@@ -0,0 +1,175 @@
+    formData.append('users[0][lastname]', lastname);
+    formData.append('users[0][password]', password);
+    formData.append('users[0][username]', username);
+    formData.append('users[0][auth]', 'oidc');
+
+    const res = await fetch(MOODLE_API_URL, {
+      method: 'POST',
+      body: formData,
+    });
+
+    const json: MoodleUser[] | MoodleException = await res.json();
+    if ('exception' in json) {
+      throw new Error(json.message);
+    }
+
+    const user: MoodleUser = json[0];
+
+    return user;
+  } catch (_) {
+    throw new Error('Failed to create the moodle user.');
+  }
+}
+
+export async function enrolMoodleUserInCourse({
+  userid,
+  courseid,
+}: {
+  userid: number;
+  courseid: number;
+}): Promise<void> {
+  try {
+    const formData = new FormData();
+    formData.append('wstoken', MOODLE_SECRET);
+    formData.append('wsfunction', 'enrol_manual_enrol_users');
+    formData.append('moodlewsrestformat', 'json');
+    formData.append('enrolments[0][roleid]', MOODLE_STUDENT_ROLE_ID.toString());
+    formData.append('enrolments[0][userid]', userid.toString());
+    formData.append('enrolments[0][courseid]', courseid.toString());
+
+    const res = await fetch(MOODLE_API_URL, {
+      method: 'POST',
+      body: formData,
+    });
+    const json: MoodleException | null = await res.json();
+
+    if (json && 'exception' in json) {
+      throw new Error(json.message);
+    }
+  } catch (_) {
+    throw new Error('Failed to enrol the moodle user into the course.');
+  }
+}
+
+export async function getMoodleCoursesByUser(
+  userid: number
+): Promise<MoodleCourse[]> {
+  try {
+    const formData = new FormData();
+    formData.append('wstoken', MOODLE_SECRET);
+    formData.append('wsfunction', 'core_enrol_get_users_courses');
+    formData.append('moodlewsrestformat', 'json');
+    formData.append('userid', userid.toString());
+
+    const res = await fetch(MOODLE_API_URL, {
+      method: 'POST',
+      body: formData,
+    });
+    const json: MoodleCourse[] | MoodleException = await res.json();
+
+    if ('exception' in json) {
+      throw new Error(json.message);
+    }
+
+    return json;
+  } catch (_) {
+    throw new Error(`Failed to fetch the user's enrolled courses.`);
+  }
+}
+
+export function getMoodleCourseUrl(courseid: number) {
+  return `${MOODLE_HOST_URL}/course/view.php?id=${courseid}`;
+}
+
+type GetMoodleCourseUrl = {
+  courseid: number;
+  userid?: number;
+};
+
+export async function createMoodleCourseUrl(
+  req: NextRequest,
+  { courseid, userid }: GetMoodleCourseUrl
+): Promise<string | undefined> {
+  const authResponse = await isAuthenticated({ req });
+
+  if (authResponse instanceof AuthErrorResponse) {
+    return;
+  }
+
+  userid = userid || (await getMoodleUserByEmail(authResponse.email))?.id;
+
+  if (!userid) {
+    return;
+  }
+
+  const courses = await getMoodleCoursesByUser(userid);
+
+  return (
+    courses.find(({ id }) => id === courseid) && getMoodleCourseUrl(courseid)
```

Comment:
I think we can create a var and then return, this will help us to identify when we add logging.